x-pack/plugins/endpoint/server/routes/resolver/utils/tree.ts

```diff
@@ -0,0 +1,243 @@
+}
+
+// This class aids in constructing a tree of process events. It works in the following way:
+//
+// 1. We construct a tree structure starting with the root node for the event we're requesting.
+// 2. We leverage the ability to pass hashes and arrays by reference to construct a fast cache of
+//    process identifiers that updates the tree structure as we push values into the cache.
+//
+// When we query a single level of results for child process events we have a flattened, sorted result
+// list that we need to add into a constructed tree. We also need to signal in an API response whether
+// or not there are more child processes events that we have not yet retrieved, and, if so, for what parent
+// process. So, at the end of our tree construction we have a relational layout of the events with no
+// pagination information for the given parent nodes. In order to actually construct both the tree and
+// insert the pagination information we basically do the following:
+//
+// 1. Using a terms aggregation query, we return an approximate roll-up of the number of child process
+//    "creation" events, this gives us an estimation of the number of associated children per parent
+// 2. We feed these child process creation event "unique identifiers" (basically a process.entity_id)
+//    into a second query to get the current state of the process via its "lifecycle" events.
+// 3. We construct the tree above with the "lifecycle" events.
+// 4. Using the terms query results, we mark each non-leaf node with the number of expected children, if our
+//    tree has less children than expected, we create a pagination cursor to indicate "we have a truncated set
+//    of values".
+// 5. We mark each leaf node (the last level of the tree we're constructing) with a "null" for the expected
+//    number of children to indicate "we have not yet attempted to get any children".
+//
+// Following this scheme, we use exactly 2 queries per level of children that we return--one for the pagination
+// and one for the lifecycle events of the processes. The downside to this is that we need to dynamically expand
+// the number of documents we can retrieve per level due to the exponential fanout of child processes,
+// what this means is that noisy neighbors for a given level may hide other child process events that occur later
+// temporally in the same level--so, while a heavily forking process might get shown, maybe the actually malicious
+// event doesn't show up in the tree at the beginning.
+export class Tree {
+  protected cache: Record<string, Node>;
+  protected root: Node;
+  protected id: string;
+
+  constructor(id: string) {
+    const root = createNode(id);
+    this.id = id;
+    this.cache = { [id]: root };
+    this.root = root;
+  }
+
+  public render() {
+    return this.root;
+  }
+
+  public ids() {
+    return Object.keys(this.cache);
+  }
+
+  public static async merge(
```

Comment:
I'd like to remove this function and have a single tree build up as we add events to it, see above.